freqtrade/freqtradebot.py

```diff
@@ -204,33 +212,46 @@ def get_target_bid(self, pair: str, tick: Dict = None) -> float:
 
         return used_rate
 
-    def _get_trade_stake_amount(self, pair) -> Optional[float]:
+    def get_trade_stake_amount(self, pair) -> Optional[float]:
         """
-        Check if stake amount can be fulfilled with the available balance
-        for the stake currency
-        :return: float: Stake Amount
+        Calculate stake amount for the trade
+        :return: float: Stake amount
         """
+        stake_amount: Optional[float]
         if self.edge:
-            return self.edge.stake_amount(
+            stake_amount = self.edge.stake_amount(
                 pair,
                 self.wallets.get_free(self.config['stake_currency']),
                 self.wallets.get_total(self.config['stake_currency']),
                 Trade.total_open_trades_stakes()
             )
         else:
             stake_amount = self.config['stake_amount']
+            if stake_amount == constants.UNLIMITED_STAKE_AMOUNT:
+                stake_amount = self._calculate_unlimited_stake_amount()
+
+        return self._check_available_stake_amount(stake_amount)
 
+    def _calculate_unlimited_stake_amount(self) -> Optional[float]:
+        """
+        Calculate stake amount for "unlimited" stake amount
+        :return: None if max number of trades reached
+        """
+        free_open_trades = self.get_free_open_trades()
+        if not free_open_trades:
+            return None
         available_amount = self.wallets.get_free(self.config['stake_currency'])
+        return available_amount / free_open_trades
 
-        if stake_amount == constants.UNLIMITED_STAKE_AMOUNT:
-            open_trades = len(Trade.get_open_trades())
-            if open_trades >= self.config['max_open_trades']:
-                logger.warning("Can't open a new trade: max number of trades is reached")
-                return None
-            return available_amount / (self.config['max_open_trades'] - open_trades)
+    def _check_available_stake_amount(self, stake_amount: Optional[float]) -> Optional[float]:
+        """
+        Check if stake amount can be fulfilled with the available balance
+        for the stake currency
+        :return: float: Stake amount
+        """
+        available_amount = self.wallets.get_free(self.config['stake_currency'])
```

Comment:
I would maybe move this to the outermost function (`get_trade_stake_amount()`) and pass the result around.

That would increase the ease at which we can test these seperately since they'd be completely selfcontained - and we'd not need to mock get_free to return the correct thing.

Reply:
Hmm, I would not like to do that, this function is now more like a blackbox, fetching internally the info it needs to check the passed stake_amount with. If it will check against relative stake amount ratio, that value should also be fetched inside this function, not passed as the parameter...